src/persist-client/src/internal/cache.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! In-process caches of [Blob].
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use moka::sync::Cache;
+use mz_ore::bytes::SegmentedBytes;
+use mz_ore::cast::CastFrom;
+use mz_persist::location::{Atomicity, Blob, BlobMetadata, ExternalError};
+use tracing::error;
+
+use crate::cfg::PersistConfig;
+use crate::internal::metrics::Metrics;
+
+// In-memory cache for [Blob].
+#[derive(Debug)]
+pub struct BlobMemCache {
+    metrics: Arc<Metrics>,
+    cache: Cache<String, SegmentedBytes>,
+    blob: Arc<dyn Blob + Send + Sync>,
+}
+
+impl BlobMemCache {
+    pub fn new(
+        cfg: &PersistConfig,
+        metrics: Arc<Metrics>,
+        blob: Arc<dyn Blob + Send + Sync>,
+    ) -> Arc<dyn Blob + Send + Sync> {
+        let cache = Cache::<String, SegmentedBytes>::builder()
+            .max_capacity(u64::cast_from(cfg.blob_cache_mem_limit_bytes))
+            .weigher(|k, v| {
+                u32::try_from(v.len()).unwrap_or_else(|_| {
+                    // We chunk off blobs at 128MiB, so the length should easily
+                    // fit in a u32.
+                    error!(
+                        "unexpectedly large blob in persist cache {} bytes: {}",
+                        v.len(),
+                        k
+                    );
+                    u32::MAX
+                })
+            })
+            .build();
```

Comment:
could be interesting to add a listener so we can track how many removals come from explicit `delete` calls vs size-based evictions

Reply:
great idea! will do